apps/evm/lib/evm/builtin.ex

```diff
@@ -219,24 +220,107 @@ defmodule EVM.Builtin do
       gas < @g_ec_mult ->
         {0, %EVM.SubState{}, exec_env, :failed}
 
-      x > IntegerModP.default_modulus() || y > IntegerModP.default_modulus() ->
+      x > FQ.default_modulus() || y > FQ.default_modulus() ->
         {0, %EVM.SubState{}, exec_env, :failed}
 
       true ->
         calculate_ec_mult({x, y}, scalar, gas, exec_env)
     end
   end
 
+  @spec ec_pairing(EVM.Gas.t(), EVM.ExecEnv.t()) ::
+          {EVM.Gas.t(), EVM.SubState.t(), EVM.ExecEnv.t(), EVM.VM.output()}
+  def ec_pairing(gas, exec_env) do
+    data = exec_env.data
+    pair_size = byte_size(exec_env.data)
+
+    gas_cost = @g_ec_pairing_point * div(pair_size, 192) + @g_ec_pairing
+
+    if rem(pair_size, 192) != 0 || gas < gas_cost do
+      {0, %EVM.SubState{}, exec_env, :failed}
+    else
+      case read_pairs(data) do
+        {:error, _} ->
+          {0, %EVM.SubState{}, exec_env, :failed}
+
+        pairs ->
+          pairing_result = pairing(pairs)
+          result = if pairing_result == FQ12.one(), do: 1, else: 0
+
+          output = Helpers.left_pad_bytes(result, 32)
+          gas_cost = @g_ec_pairing_point * div(pair_size, 192) + @g_ec_pairing
+
+          {gas - gas_cost, %EVM.SubState{}, exec_env, output}
+      end
+    end
+  end
+
+  @spec read_pairs(binary()) ::
+          {:ok, [{{FQ.t(), FQ.t()}, {FQP.t(), FQP.t()}}]} | {:error, String.t()}
```

Comment:
These points seem to come in pairs (x and y I imagine). Do you think it would make sense to extract a type for that? Maybe something like `FQ.point()`? Same thing with `FQP.point()`?